File: practice/mergeSort.js
arr = [5, 2, 4, 6, 1, 3]

// Merge Sort Implentation (Recursion)
function mergeSort (unsortedArray) {
    // No need to sort the array if the array only has one element or empty
    if (unsortedArray.length <= 1) {
      return unsortedArray;
    }
    // In order to divide the array in half, we need to figure out the middle
    const middle = Math.floor(unsortedArray.length / 2);
  
    // This is where we will be dividing the array into left and right
    const left = unsortedArray.slice(0, middle);
    const right = unsortedArray.slice(middle);
  
    // Using recursion to combine the left and right
    return merge(
      mergeSort(left), mergeSort(right)
    );
  }

  // Merge the two arrays: left and right
function merge (left, right) {
    let resultArray = [], leftIndex = 0, rightIndex = 0;
  
    // We will concatenate values into the resultArray in order
    while (leftIndex < left.length && rightIndex < right.length) {
      if (left[leftIndex] < right[rightIndex]) {
        resultArray.push(left[leftIndex]);
        leftIndex++; // move left array cursor
      } else {
        resultArray.push(right[rightIndex]);
        rightIndex++; // move right array cursor
      }
    }

    // Check for elements remaining from either left OR the right
    while (leftIndex < left.length) {
          resultArray.push(left[leftIndex]);
          leftIndex++; // move left array cursor
      }
    while (rightIndex < right.length) {
        resultArray.push(right[rightIndex]);
        rightIndex++; // move left array cursor
    }
  
    return resultArray
  }


result = mergeSort(arr)
console.log("mergeSort result:", result);

File: practice/merge_sort.py
arr = [5, 2, 4, 6, 1, 3]

def merge_sort(A):
    if len(A) <= 1:
        return A

    # finding the mid point
    mid = len(A) // 2

    # dividing into 2 parts
    L = A[:mid]
    R = A[mid:]

    # using recursion to combine the left and right
    return merge(A, merge_sort(L), merge_sort(R))
    
    

def merge(A, L, R):

    i = j = k = 0

    # merging L and R arrays in order
    while i < len(L) and j < len(R):
        if L[i] < R[j]:
            A[k] = L[i]
            i += 1
        else:
            A[k] = R[j]
            j += 1
        k += 1
    
    # checking if any element was left
    while i < len(L):
        A[k] = L[i]
        i += 1
        k += 1
    while j < len(R):
        A[k] = R[j]
        j += 1
        k += 1

    return A

    

if __name__ == '__main__':
    merge_sort_result = merge_sort(arr)
    print('merge_sort_result:', merge_sort_result)
